File: js/src/interiorPointPoint.ts
import type { Geometry } from "geojson";

import { getCentroid } from "./centroid";
import type { Coordinate } from "./geometryAdapter";
import { distance, isGeometryEmpty } from "./geometryAdapter";

/**
 * Computes a point in the interior of an point geometry.
 * <h2>Algorithm</h2>
 * Find a point which is closest to the centroid of the geometry.
 *
 * @jts InteriorPointPoint
 */
export class InteriorPointPoint {
  private centroid: Coordinate | null;
  private minDistance = Number.MAX_VALUE;
  private interiorPoint: Coordinate | null = null;

  /** @jts InteriorPointPoint#InteriorPointPoint(Geometry) */
  constructor(g: Geometry) {
    this.centroid = getCentroid(g);
    this.addGeometry(g);
  }

  /**
   * Tests the point(s) defined by a Geometry for the best inside point.
   * If a Geometry is not of dimension 0 it is not tested.
   *
   * @param geom the geometry to add
   * @jts InteriorPointPoint#add(Geometry)
   */
  private addGeometry(geom: Geometry): void {
    if (isGeometryEmpty(geom)) return;

    switch (geom.type) {
      case "Point":
        this.addCoordinate(geom.coordinates);
        break;
      // JTS's MultiPoint is a GeometryCollection and falls through to the
      // collection branch there; GeoJSON's is not, so it is expanded here.
      case "MultiPoint":
        for (const c of geom.coordinates) this.addCoordinate(c);
        break;
      case "GeometryCollection":
        for (const g of geom.geometries) this.addGeometry(g);
        break;
      default:
        break;
    }
  }

  /** @jts InteriorPointPoint#add(Coordinate) */
  private addCoordinate(point: Coordinate): void {
    // `centroid` is null only for an empty input, which returns from
    // addGeometry before this is reachable.
    const dist = distance(point, this.centroid as Coordinate);
    if (dist < this.minDistance) {
      this.interiorPoint = [...point];
      this.minDistance = dist;
    }
  }

  /**
   * Gets the computed interior point.
   *
   * @return the computed interior point, or null if the input geometry is empty
   * @jts InteriorPointPoint#getInteriorPoint()
   */
  getInteriorPoint(): Coordinate | null {
    return this.interiorPoint;
  }
}

/**
 * Computes an interior point for the puntal components of a Geometry.
 *
 * @param geom the geometry to compute
 * @return the computed interior point, or null if the geometry has no puntal components
 * @jts InteriorPointPoint#getInteriorPoint(Geometry)
 * @jts-deviate module-level name — `getInteriorPoint` would collide with the
 *   same static factory in the other three modules.
 */
export function interiorPointPoint(geom: Geometry): Coordinate | null {
  const intPt = new InteriorPointPoint(geom);
  return intPt.getInteriorPoint();
}

File: rs/core/src/algorithm/interior_point_point.rs
//! Computes a point in the interior of an point geometry.
//!
//! # Algorithm
//!
//! Find a point which is closest to the centroid of the geometry.
//!
//! @jts InteriorPointPoint

use geo_types::{Coord, Geometry};

use crate::algorithm::centroid::get_centroid;
use crate::geometry_adapter::{distance, is_geometry_empty};

pub(crate) struct InteriorPointPoint {
    centroid: Option<Coord<f64>>,
    min_distance: f64,
    interior_point: Option<Coord<f64>>,
}

impl InteriorPointPoint {
    /// @jts InteriorPointPoint#InteriorPointPoint(Geometry)
    pub(crate) fn new(g: &Geometry<f64>) -> Self {
        let mut int_pt = Self {
            centroid: get_centroid(g),
            min_distance: f64::MAX,
            interior_point: None,
        };
        int_pt.add_geometry(g);
        int_pt
    }

    /// Tests the point(s) defined by a Geometry for the best inside point.
    /// If a Geometry is not of dimension 0 it is not tested.
    ///
    /// @jts InteriorPointPoint#add(Geometry)
    fn add_geometry(&mut self, geom: &Geometry<f64>) {
        if is_geometry_empty(geom) {
            return;
        }
        match geom {
            Geometry::Point(p) => self.add_coordinate(p.0),
            // JTS's MultiPoint is a GeometryCollection and falls through to the
            // collection branch there; geo-types' is not, so it is expanded here.
            Geometry::MultiPoint(mp) => {
                for p in &mp.0 {
                    self.add_coordinate(p.0);
                }
            }
            Geometry::GeometryCollection(gc) => {
                for g in &gc.0 {
                    self.add_geometry(g);
                }
            }
            _ => {}
        }
    }

    /// @jts InteriorPointPoint#add(Coordinate)
    fn add_coordinate(&mut self, point: Coord<f64>) {
        // `centroid` is None only for an empty input, which returns from
        // `add_geometry` before this is reachable.
        let dist = distance(
            point,
            self.centroid
                .expect("centroid is set for a non-empty input"),
        );
        if dist < self.min_distance {
            self.interior_point = Some(point);
            self.min_distance = dist;
        }
    }

    /// Gets the computed interior point.
    ///
    /// Returns the computed interior point, or `None` if the input geometry is empty.
    ///
    /// @jts InteriorPointPoint#getInteriorPoint()
    pub(crate) fn get_interior_point(&self) -> Option<Coord<f64>> {
        self.interior_point
    }
}

/// Computes an interior point for the puntal components of a Geometry.
///
/// Returns the computed interior point, or `None` if the geometry has no
/// puntal components.
///
/// @jts InteriorPointPoint#getInteriorPoint(Geometry)
/// @jts-deviate module-level name — `get_interior_point` would collide with the
///   same static factory in the other three modules.
pub(crate) fn interior_point_point(geom: &Geometry<f64>) -> Option<Coord<f64>> {
    let int_pt = InteriorPointPoint::new(geom);
    int_pt.get_interior_point()
}
